kernel/Utilities/SparsityEstimator.cpp

```diff
@@ -15,310 +18,416 @@ SparsityEstimator::SparsityEstimator(const GlobalIndexing& rowindexing,
-        nonZeroPerRowOwned.clear();
-        nonZeroPerRowOwned.resize(totalNumberOfDoF, 0);
-        nonZeroPerRowNonOwned.clear();
-        nonZeroPerRowNonOwned.resize(totalNumberOfDoF, 0);
+
+    nonZeroPerRowOwned.assign(totalNumberOfDoF, 0);
+    nonZeroPerRowNonOwned.assign(totalNumberOfDoF, 0);
+
+    if (rowIndexing_.getMesh() == nullptr || totalNumberOfDoF == 0) {
+        // Nothing to do, result has already been resized.
         return;
     }
 
-    Workspace workspace;
-
-    // Resize and initialize with error data
-    nonZeroPerRowOwned.assign(totalNumberOfDoF, -1);
-    nonZeroPerRowNonOwned.assign(totalNumberOfDoF, -1);
-
-    // To compute the number of non zero's in each row we assume
-    // (pessimistically) that all face matrices are filled. For each DoF we then
-    // compute the set of other DoFs that contribute to at least one face
-    // matrix, and counting these gives the number of non zeros in the row.
-    //
-    // Counting DoFs is done by going over all elements where a basis function
-    // has support. All basis functions with support on these elements can
-    // result in non zero element matrix entries. For non zero face matrix
-    // entries we also need to consider all basis functions that have support on
-    // the element adjacent to the elements where the basis function has
-    // support.
-    //
-    // For efficiency purposes we do not traverse each basis function
-    // separately, instead we compute it for all basis functions of a single
-    // element/face/etc.
+    Detail::LocalMatrixSideDoFs rowDoFs(rowIndexing_, Detail::ROWS);
+    Detail::LocalMatrixSideDoFs columnDoFs(columnIndexing_, Detail::COLUMNS);
+
+    std::map<Detail::DoFInterval, Detail::DoFIntervals> owned;
+    std::map<Detail::DoFInterval, Detail::DoFIntervals> nonOwned;
 
     for (const Base::Element* element :
-         rowIndexing_.getMesh()->getElementsList()) {
-        if (!hasLocalDoFs(element, rowIndexing_.getIncludedUnknowns())) {
-            // If there are no DoFs associated with this element, then then
-            // there are no rows in the matrix for this element and there is
-            // nothing to count.
-            continue;
-        }
-        workspace.clear();
-        // Add local basis functions
-        addElementDoFs(element, workspace);
-
-        // Face matrix coupling with the element on the other side of the face
-        if (includeFaceCoupling) {
-            for (const Base::Face* face : element->getFacesList()) {
-                if (!face->isInternal()) continue;
-                const Base::Element* other = face->getPtrOtherElement(element);
-                addElementDoFs(other, workspace);
-            }
-        }
+         rowIndexing_.getMesh()->getElementsList(Base::IteratorType::GLOBAL)) {
+        // Face, Edge and Node based DoFs will be used in the element matrices
+        // on the ghost layer. So we need to iterate over all elements and not
```

Comment:
what is the ghost layer?

Reply:
In an MPI setting, where you have multiple processes, you distribute the mesh so that each process owns part of the mesh. This ownership then results in a distribution of the work, for example, to assemble a global matrix as combination of element  and face matrices, each process only adds the element matrices for the elements and faces it owns.

Ideally, each process would then load only the part of the mesh that you own. Thereby saving the memory, time, etc. of not loading the mesh parts owned by the other tens of processors. However, only knowing about the mesh the process owns is insufficient to do all its calculations. For example, when constructing a face matrix on the subdomain boundary (=boundary between the ownership region of two processes), you would need to know about the element that is owned by the other process. 

So each processor also loads a small layer of elements that surround the elements that it owns itself. This layer of elements, in our case 1 element wide, is known as the ghost layer. The hard thing with these elements (faces, edges and nodes) is that the information on them may be incomplete. For example, a face in the ghost layer may only have an element on one side because the other element is not loaded on the process.

Reply:
Ah okay. Thanks.